### PR TITLE
[v6r15] Fixed a concurrency issue in the aggregation of emails

### PR DESCRIPTION
**Pull request by vipersec**
-  The previous implementation had a concurrency issue when many instances were trying to write to the `json` file **simultaneously**. I fixed this by using `SQLite`.

- If the environmental variable `DIRAC` does not exists, it is going to use the current directory (This was failing in the machine which runs `Travis CI` because it doesn't have the `DIRAC` environmental variable) 

- I removed the EmailAction test that I had written since it is invalid now.

### CODE REVIEW

File: ResourceStatusSystem/Agent/EmailAgent.py [hunk 1]
```diff
@@ -25,8 +25,11 @@ def __init__( self, *args, **kwargs ):
     AgentModule.__init__( self, *args, **kwargs )
     self.diracAdmin = None
     self.default_value = None
-    self.cacheFile = None
-    self.cacheFile = os.getenv('DIRAC') + 'work/ResourceStatus/cache.json'
+
+    if os.getenv('DIRAC'):
```
Comment (chaen) [inline ResourceStatusSystem/Agent/EmailAgent.py]: ```if DIRAC in os.environ```

File: ResourceStatusSystem/Agent/EmailAgent.py [hunk 2]
```diff
@@ -25,8 +25,11 @@ def __init__( self, *args, **kwargs ):
...
+      self.cacheFile = os.getenv('DIRAC') + '/work/ResourceStatus/cache.db'
```
Comment (chaen) [inline ResourceStatusSystem/Agent/EmailAgent.py]: you can use ```os.path.join``` which is safer

File: ResourceStatusSystem/Agent/EmailAgent.py [hunk 3]
```diff
@@ -25,8 +25,11 @@ def __init__( self, *args, **kwargs ):
...
+    else:
+      self.cacheFile = 'cache.db'
```
Comment (chaen) [inline ResourceStatusSystem/Agent/EmailAgent.py]: Here you could use ```os.path.realpath```, so ```self.cacheFile``` contains a full path

File: ResourceStatusSystem/Agent/EmailAgent.py [hunk 4]
```diff
@@ -38,55 +41,25 @@ def initialize( self ):
...
-      self.cacheFile += ".send"
+    result = conn.execute("SELECT DISTINCT SiteName from ResourceStatusCache;")
+    for site in result:
+      cursor = conn.execute("SELECT StatusType, ResourceName, Status, Time, PreviousStatus from ResourceStatusCache WHERE SiteName='"+ site[0] +"';")
 
-      #load the file
-      with open(self.cacheFile, 'r') as f:
-        new_dict = json.load(f)
+      email_body = ""
+      for row in cursor:
```
Comment (chaen) [inline ResourceStatusSystem/Agent/EmailAgent.py]: it is maybe more readable to unpack the values. For example 
``` for x, y, z in cursor```

This way, we know what is the meaning of ```row[0], row[1]...```

File: ResourceStatusSystem/PolicySystem/Actions/EmailAction.py [hunk 5]
```diff
@@ -24,7 +21,10 @@ def __init__( self, name, decisionParams, enforcementResult, singlePolicyResults
     super( EmailAction, self ).__init__( name, decisionParams, enforcementResult,
                                          singlePolicyResults, clients )
 
-    self.cacheFile = os.getenv('DIRAC') + 'work/ResourceStatus/' + 'cache.json'
+    if os.getenv('DIRAC'):
```
Comment (chaen) [inline ResourceStatusSystem/PolicySystem/Actions/EmailAction.py]: same as above

File: ResourceStatusSystem/PolicySystem/Actions/EmailAction.py [hunk 6]
```diff
@@ -55,47 +55,35 @@ def run( self ):
...
       return S_ERROR( 'reason should not be None' )
 
-    siteName = getSitesForSE(name)['Value'][0]
-    time     = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
-    dict    = { 'name': name, 'statusType': statusType, 'status': status, 'time': time, 'previousStatus': previousStatus }
+    siteName = getSitesForSE(name)
 
-    actionResult = self._addtoJSON(siteName, dict)
+    if not siteName['OK']:
+      siteName = "Unassigned Resources"
```
Comment (chaen) [inline ResourceStatusSystem/PolicySystem/Actions/EmailAction.py]: you might want to log the error

**Comment by chaen** [comment]
Why is the test invalid now ? You can replace it with a new one if needed ;-)

File: ResourceStatusSystem/Agent/EmailAgent.py (hunk 1 above)
Comment (vipersec) [inline ResourceStatusSystem/Agent/EmailAgent.py]: why is it prefered? If there is an environmental variable named `DIRACwork` (maybe in the future) then the `if DIRAC in os.environ` will return `True` and the `os.getenv('DIRAC')` will fail, right?

Comment (chaen) [inline ResourceStatusSystem/Agent/EmailAgent.py]: ```os.environ``` is a dict, so no reason for it to behave like string :-)

Just to note, however, there is a trick with tuples:
```
In [14]: t1 = ( 'something', 'else')

In [15]: t2 = ( 'something' )

In [16]: 'some' in t1
Out[16]: False

In [17]: 'some' in t2
Out[17]: True
```

If you have a single element in your tuple, without the ```,``` in it in the end, it is iterpreted as the type inside this tuple, and then you can fall in the pitfal you mention

Comment (vipersec) [inline ResourceStatusSystem/Agent/EmailAgent.py]: Thanks @chaen I didn't know that :)

File: ResourceStatusSystem/Agent/EmailAgent.py [hunk 7]
```diff
@@ -38,55 +41,25 @@ def initialize( self ):
 
   def execute( self ):
 
-    #if the file exists and it is not empty
-    if os.path.isfile(self.cacheFile) and os.stat(self.cacheFile).st_size:
+    conn = sqlite3.connect(self.cacheFile)
```
Comment (fstagni) [inline ResourceStatusSystem/Agent/EmailAgent.py]: I suggest to use the `with` statement to interact with sqlite3: the code will be more compact and safe: `with sqlite3.connect(self.cacheFile) as conn` and then you don't need to invoke `conn.close()`

**Comment by fstagni** [comment]
As @chaen says: you'll need a new test now. All the rest seems OK for me.

File: ResourceStatusSystem/PolicySystem/Actions/EmailAction.py (hunk 6 above)
Comment (vipersec) [inline ResourceStatusSystem/PolicySystem/Actions/EmailAction.py]: Done https://github.com/DIRACGrid/DIRAC/pull/2899/commits/be7cd02c950afe3503ebf3dc8657e2ea78824677

**Comment by fstagni** [comment]
Review OK

**Pull request merged**